### PR TITLE
Reduce logs in security/pkg/nodeagent/sds/

### PR DESCRIPTION
**Pull request by lei-tang**
https://github.com/istio/istio/issues/13033

### CODE REVIEW

**Comment by howardjohn** [comment]
This doesn't fix the issue, all of the error messages are still there. I guess it cuts it in half though

**Comment by howardjohn** [comment]
What happens if `.Serve()` never succeeds? There is going to be millions of logs produced?

**Comment by lei-tang** [comment]
Error messages are useful for debugging purpose.  If .Serve() never succeeds, the error message will be output consistently and consecutively (instead of drown out among other logs), which highlights the Serve() problem. If a user does not want to see verbose logs, the user may choose different log levels, or pipe logs through filters to filter out the logs that the user are not interested in.

**Comment by howardjohn** [comment]
> Error messages are useful for debugging purpose. If .Serve() never succeeds, the error message will be output consistently and consecutively (instead of drown out among other logs), which highlights the Serve() problem. If a user does not want to see verbose logs, the user may choose different log levels, or pipe logs through filters to filter out the logs that the user are not interested in.

Getting 1 million of the same error message is not helpful for debugging purposes. If making it "not drown out among other logs" was a good idea, then should we have every error message log 1000 times so they don't get drowned out? These are logged at the error level, so changing the log level isn't really a good idea...

These are logged in passing unit tests. A **passing** unit test should not log 100k error messages....

**Comment by howardjohn** [comment]
This logs 100k times in a single second, its not like it logs this once per second

**Comment by wattli** [comment]
Agree with Howard's point, i think it makes sense to have exponential + threshold, i.e., if the server can never be brought up, we should shut down the instance to signal error message.

**Comment by lei-tang** [comment]
@howardjohn "This logs 100k times in a single second", which log is outputed 100k times in a single second? If so, this log should be removed due to 1) outputed 100k times in a single second decreases the program performance. 2) makes it difficult to view other logs.

**Comment by howardjohn** [comment]
Just run `go test ./security/pkg/nodeagent/sds/...` or see example here: https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/12872/istio-unit-tests/23157

**Comment by lei-tang** [comment]
Added counters to limit the max times error messages can be outputed. Please take a look. If this counting mechanism is not enough, a more complicated backoff mechanism can be devised and added.

File: security/pkg/nodeagent/sds/server.go [hunk 1]
```diff
@@ -193,15 +202,23 @@ func (s *Server) initGatewaySdsService(options *Options) error {
 	}
 
 	go func() {
+		log.Info("Start SDS grpc server for ingress gateway proxy")
+		numErrorOutputTime1 := 0
```
Comment (incfly) [inline security/pkg/nodeagent/sds/server.go]: Time1 and Time2 are not good name for var.

File: security/pkg/nodeagent/sds/server.go [hunk 2]
```diff
@@ -193,15 +202,23 @@ func (s *Server) initGatewaySdsService(options *Options) error {
 	}
 
 	go func() {
+		log.Info("Start SDS grpc server for ingress gateway proxy")
```
Comment (incfly) [inline security/pkg/nodeagent/sds/server.go]: Please change to expotential back off, which is really simple. Psedu code.

```go
waitTime := time.Second
for i := 0; i < 10; i++ {
  log.Errorf("Failed to start gRPC server for SDS, retry in %v", waitTime)
  time.Sleep(waitTime)
  waitTime *= 2
}
```

File: security/pkg/nodeagent/sds/server.go [hunk 3]
```diff
@@ -165,15 +166,23 @@ func (s *Server) initWorkloadSdsService(options *Options) error { //nolint: unpa
...
 	go func() {
+		log.Info("Start SDS grpc server")
+		numErrorOutputTime1 := 0
+		numErrorOutputTime2 := 0
 		for {
 			// Retry if Serve() fails
-			log.Info("Start SDS grpc server")
 			if err = s.grpcWorkloadServer.Serve(s.grpcWorkloadListener); err != nil {
-				log.Errorf("SDS grpc server for workload proxies failed to start: %v", err)
+				if numErrorOutputTime1 < maxErrorMessageOutputTimes {
```
Comment (incfly) [inline security/pkg/nodeagent/sds/server.go]: What's the typical reason retry would help? If it's host file mount readiness issue, sleep can help here.

File: security/pkg/nodeagent/sds/server.go (hunk 2 above)
Comment (lei-tang) [reply security/pkg/nodeagent/sds/server.go]: Done: the PR has been revised accordingly.

File: security/pkg/nodeagent/sds/server.go (hunk 3 above)
Comment (lei-tang) [reply security/pkg/nodeagent/sds/server.go]: Sleeping mechanism is added.

File: security/pkg/nodeagent/sds/server.go (hunk 1 above)
Comment (lei-tang) [reply security/pkg/nodeagent/sds/server.go]: Done: revised based on the comments.

**Comment by incfly** [comment]
Thanks @lei-tang for the fix.

> with this change, we need a way to stop nodeagent when exit go func(){}, so that k8s could restart it.
otherwise nodeagent will hang there after hitting maxRetryTimes in real env. We could have a different PR for this.

@quanjielin  Not sure what you mean. In this current code, I expect when maxRetry hit, node agent will finish complete automatically. Or you want to explicit use another os.Exit(-1) non zero return code, which is necessary for k8s to restart a pod?

By the way, I don't see why we need to use a goroutine for `server.Serve()` though. It's invokved from default main goroutine, main.go, `sds.NewServer`. We can just let it hanging there using default goroutine. And return error if needed.

Thoughts?

**Comment by lei-tang** [comment]
https://github.com/istio/istio/issues/13439

**Pull request merged**